### ISSUE TITLE
Getting a zero gradient from autograd after vectorising function

### ISSUE THREAD

**Issue by apal90** [google/jax#1407]
This post concerns a new problem that emerged from solving an issue I had opened here: https://github.com/google/jax/issues/1318.

I have a function, whose gradient I am trying to calculate using autograd. I had previously used `autograd.grad()` with success, but I need to speed this up, hence I turned to `jax,grad()`. Initially (see the above link), the function was not properly vectorised (there were many loops instead of vector operations), and hence `jax.grad()` was taking infinitely long to calculate the gradient, and often even killing the kernel.

**PROBLEM**: 
After vectorising, I no longer get this problem. The gradient calculates smoothly, but instead, I get a zero gradient, using both `autograd.grad()` and `jax.grad()`! The scalar returned by the function, however, is still the same. So, something happened during the vectorisation procedure that prevents both autograd and jax from calculating the true gradient. Can someone please help me figure out what?

**EXAMPLE CODE:**
Here is a colab notebook containing a self-sufficient example: https://colab.research.google.com/drive/1VCPiRTLfxOflooDtTlqHI4VzYQkXzaUn 

I have put in three scenarios.

1. The original code (pre-vectorisation), with gradient calculated using `autograd.grad`. This gives values for the scalar that I trust, and also a gradient that I feel that I can trust. At least, it's non-zero.
2. Vectorised code, using `autograd.numpy` and `autograd.grad`. I get back the same scalar values, but a zero gradient.
3. Vectorised code, using `jax` functions (`jax.numpy`, `jax.vmap`, `jax.grad`, etc.). Again, I get back the same scalar values, but a zero gradient.

**Comment by NickRothbacher** [comment]
Any update on this issue? I'm getting similar behavior after vectorizing my model.

**Comment by hawkinsp** [comment]
I'm sorry we didn't get a chance to look at this issue before now. There are two bugs in your code.

The first is that `jax.ops.index_update` returns a new copy of the array, rather than updating in-place. See https://jax.readthedocs.io/en/latest/_autosummary/jax.ops.index_update.html#jax.ops.index_update

So you must write something like:
```
    curvVector = index_update(curvVector, 1, aVector[4]) # c12 = aVector[4]
```

The second is that in `getCurvatures` you wrote `xyzInit` where you probably meant `xyz`:

```
partialFunc = partial(getCurvature_triangle, xyzInit) 
```
vs.
```
partialFunc = partial(getCurvature_triangle, xyz)
```
This bug means that there is no dependence of the `xyz` input on the output, so the gradient indeed is `0`.

With these two issues fixed, a non-zero gradient is produced.

Hope that helps!
